### PR TITLE
Add pagerduty output support

### LINKED ISSUE #163: Suppor PagerDuty as output
**Motivation**

In my company, we use PagerDuty as a pager system.
If there are high-level priority detection, I want to get a call.

**Feature**

Page Pagerduty team if the output of Falco is higher than the minimunpriority.

**Alternatives**

**Additional context**

### PR DESCRIPTION
**Pull request by KeisukeYamashita**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area config

/area outputs

> /area tests

**What this PR does / why we need it**:

Many company uses PagerDuty and we want to create a incident on Falco reports.

**Which issue(s) this PR fixes**:

Fixes #163 

**Special notes for your reviewer**:

These are documents that might help you review:

* [Incident Creation API, PagerDuty Developers](https://developer.pagerduty.com/docs/rest-api-v2/incident-creation-api/)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -266,6 +267,13 @@ kafka:
   topic: "" # Name of the topic, if not empty, Kafka output is enabled
   # partition: 0 # Partition number of the topic.
   # minimumpriority: "debug" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+
+pagerduty:
+  # apikey: # Pagerduty API Key, if not empty, Pagerduty is enabled
+  service: "" # Service to create an incident
```
Comment (Issif) [inline README.md:273]:
```suggestion
  service: "" # Service to create an incident (mandatory)
```

File: README.md [hunk 2]
```diff
@@ -266,6 +267,13 @@ kafka:
   topic: "" # Name of the topic, if not empty, Kafka output is enabled
   # partition: 0 # Partition number of the topic.
   # minimumpriority: "debug" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+
+pagerduty:
+  # apikey: # Pagerduty API Key, if not empty, Pagerduty is enabled
```
Comment (Issif) [inline README.md:272]:
```suggestion
  # apikey: # Pagerduty API Key, if not empty, Pagerduty output is enabled
```

File: README.md [hunk 3]
```diff
@@ -384,6 +392,11 @@ The *env vars* "match" field names in *yaml file with this structure (**take car
 * **KAFKA_TOPIC**: The name of the Kafka topic
 * **KAFKA_PARTITION**: The number of the Kafka partition
 * **KAFKA_MINIMUMPRIORITY**: minimum priority of event for using this output, order is `emergency|alert|critical|error|warning|notice|informational|debug or "" (default)`
+* **PAGERDUTY_APIKEY**: Pagerduty API Key, if not empty, Pagerduty is *enabled*
```
Comment (Issif) [inline README.md:395]:
```suggestion
* **PAGERDUTY_APIKEY**: Pagerduty API Key, if not empty, Pagerduty output is *enabled*
```

File: README.md [hunk 4]
```diff
@@ -384,6 +392,11 @@ The *env vars* "match" field names in *yaml file with this structure (**take car
...
+* **PAGERDUTY_SERVICE**: Service to create an incident
```
Comment (Issif) [inline README.md:396]:
```suggestion
* **PAGERDUTY_SERVICE**: Service to create an incident (mandatory)
```

File: outputs/pagerduty.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	"log"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/PagerDuty/go-pagerduty"
+	"github.com/falcosecurity/falcosidekick/types"
+)
+
+// NewPagerdutyClient returns a new output.Client for accessing the Pagerduty API.
+func NewPagerdutyClient(config *types.Configuration, stats *types.Statistics, promStats *types.PromStatistics, statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
+	if len(config.Pagerduty.Assignee) > 0 && config.Pagerduty.EscalationPolicy != "" {
```
Comment (Issif) [inline outputs/pagerduty.go:14]: I prefer we check this at creation of client.

File: outputs/pagerduty.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+			ID:   policy,
+			Type: "escalation_policy_reference",
+		}
+	}
+
+	if _, err := c.PagerdutyClient.CreateIncident("falcosidekick", opts); err != nil {
+		go c.CountMetric(Outputs, 1, []string{"output:pagerduty", "status:error"})
+		c.Stats.Pagerduty.Add(Error, 1)
+		c.PromStats.Outputs.With(map[string]string{"destination": "pagerduty", "status": Error}).Inc()
+		log.Printf("[ERROR] : Pagerduty - %v\n", err)
```
Comment (Issif) [inline outputs/pagerduty.go:66]:
```suggestion
		log.Printf("[ERROR] : PagerDuty - %v\n", err)
```

File: outputs/pagerduty.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		return
+	}
+
+	go c.CountMetric(Outputs, 1, []string{"output:pagerduty", "status:ok"})
+	c.Stats.Pagerduty.Add(OK, 1)
+	c.PromStats.Outputs.With(map[string]string{"destination": "pagerduty", "status": OK}).Inc()
+	log.Printf("[INFO] : Pagerduty - Publish OK\n")
```
Comment (Issif) [inline outputs/pagerduty.go:73]:
```suggestion
	log.Printf("[INFO] : Pagerduty - Post OK\n")
```

File: outputs/pagerduty.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+		Stats:           stats,
+		PromStats:       promStats,
+		StatsdClient:    statsdClient,
+		DogstatsdClient: dogstatsdClient,
+		PagerdutyClient: pagerduty.NewClient(config.Pagerduty.APIKey),
+	}, nil
+}
+
+// PagerdutyPost posts incident to Pagerduty
+func (c *Client) PagerdutyPost(falcopayload types.FalcoPayload) {
```
Comment (Issif) [inline outputs/pagerduty.go:30]:
```suggestion
func (c *Client) PagerdutyCreateIncident(falcopayload types.FalcoPayload) {
```

For following their SDK

File: outputs/pagerduty.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+		Config:          config,
+		Stats:           stats,
+		PromStats:       promStats,
+		StatsdClient:    statsdClient,
+		DogstatsdClient: dogstatsdClient,
+		PagerdutyClient: pagerduty.NewClient(config.Pagerduty.APIKey),
+	}, nil
+}
+
+// PagerdutyPost posts incident to Pagerduty
```
Comment (Issif) [inline outputs/pagerduty.go:29]:
```suggestion
// PagerdutyCreateIncident posts incident to Pagerduty
```

File: handlers.go [hunk 10]
```diff
@@ -219,4 +219,8 @@ func forwardEvent(falcopayload types.FalcoPayload) {
 	if config.Kafka.HostPort != "" && (priorityMap[strings.ToLower(falcopayload.Priority)] >= priorityMap[strings.ToLower(config.Kafka.MinimumPriority)] || falcopayload.Rule == TestRule) {
 		go kafkaClient.KafkaProduce(falcopayload)
 	}
+
+	if config.Pagerduty.APIKey != "" && (priorityMap[strings.ToLower(falcopayload.Priority)] >= priorityMap[strings.ToLower(config.Pagerduty.MinimumPriority)] || falcopayload.Rule == TestRule) {
+		go kafkaClient.PagerdutyPost(falcopayload)
```
Comment (Issif) [inline handlers.go:224]:
```suggestion
		go kafkaClient.PagerdutyCreateIncident(falcopayload)
```

**Review by Issif — CHANGES_REQUESTED** [review]

File: outputs/pagerduty.go (hunk 5 above)
Comment (KeisukeYamashita) [reply outputs/pagerduty.go:14]: Do you mean check this in here? https://github.com/falcosecurity/falcosidekick/blob/add-pagerduty-support/main.go#L320-L327

File: outputs/pagerduty.go [hunk 11]
```diff
@@ -63,12 +63,12 @@ func (c *Client) PagerdutyPost(falcopayload types.FalcoPayload) {
...
-		log.Printf("[ERROR] : Pagerduty - %v\n", err)
+		log.Printf("[ERROR] : PagerDuty - %v\n", err)
 		return
 	}
 
 	go c.CountMetric(Outputs, 1, []string{"output:pagerduty", "status:ok"})
 	c.Stats.Pagerduty.Add(OK, 1)
 	c.PromStats.Outputs.With(map[string]string{"destination": "pagerduty", "status": OK}).Inc()
-	log.Printf("[INFO] : Pagerduty - Publish OK\n")
+	log.Printf("[INFO] : Pagerduty - Post OK\n")
```
Comment (Issif) [inline outputs/pagerduty.go:73]:
```suggestion
	log.Printf("[INFO] : Pagerduty - Create Incident OK\n")
```

**Review by Issif — CHANGES_REQUESTED** [review]

**Review by Issif — APPROVED** [review]
👍

**Comment by KeisukeYamashita** [comment]
Thank you! I will prepare chart update tomorrow.

**Comment by Issif** [comment]
@KeisukeYamashita I just noticed, all your commits need DCO to be approved.

**Comment by KeisukeYamashita** [comment]
@Issif  Thank you, I justed signed my commits and CI are passing 🎸

**Review by Issif — APPROVED** [review]
awesome !!

**Pull request merged**